### ISSUE TITLE
Keto Engine should support multiple subjects

### ISSUE THREAD

**Issue by mvanderlee** [ory/oathkeeper#269]
**Is your feature request related to a problem? Please describe.**

An incoming JWT might have a subject, and a list of roles. To check if a the token is authorized I'd like to be able to check all against the Keto policies.

Example:
JWT: 
```json
{
  "sub": "test@ory.com",
  "roles": [
      "role1",
      "role2"
  ]
}
```
Keto policies (condensed):
```yaml
- subjects: ["test@ory.com"]
  resources: ["/api/reports/50"]

- subjects: ["role1"]
  resources: ["/api/dashboards/daily/*"]

- subjects: ["role2"]
  resources: ["/api/dashboards/monthly/*"]
``` 

Oathkeeper authorizers:
```yaml
- handler: "keto_engine_acp_ory"
  config:
    required_resource: "/api/reports/$1"
    subject: [
      "{{.Subject}}",
      "{{.Extra.roles}}"
    ]

OR

- handler: "keto_engine_acp_ory"
  config:
    required_resource: "/api/dashboards/$1"
    subject: "{{.Subject}}|{{range $i,$a := .Extra.roles}}{{if gt $i 0 }}|{{end}}{{.}}{{end}}"
    subjectDelimiter: "|"
```

subjectDelimiter might work better? I have no idea how to implement it otherwise.

**Describe the solution you'd like**

I'd like Oathkeeper to iterate over the allowed subjects to check if any of them is authorized.
If any of them have an explicit DENY, then that should take precedence.

**Describe alternatives you've considered**

* Forego roles and only assign policies to a user.
* Have a middleman management service that interpret roles and assigns all role policies to the user directly
* Multiple authorizers with `role[$idx]`. But that won't work unless there is a max number of roles a user can be assigned. Even then, it's tedious.
* Forego the Keto engine entirely. Use the `allow` authorizer, and a custom `hydrator` mutator to a track roles independently from the ORY system, and assign hydrated claims to an id_token, that will be passed to the final service which will perform authorization itself based on said claims. 

**Additional context**

**Comment by aeneasr** [comment]
This might be tagential, but why wouldn't the roles/user assignment be tracked as part of Keto itself but instead as part of the token?

**Comment by MichielVanderlee** [comment]
Well, I've based this on the current acp-ory engine's documentation: https://www.ory.sh/docs/keto/engines/acp-ory#roles
Which works, as long as you only have a single role. When you assign multiple roles it no longer works. 

One thing to keep in mind as well is that a role might be used for more things than just the ORY ecosystem.
Take AWS Cognito, which I'm using, for example. It calls them 'groups' and allows me to assign an IAM roles to a group as well. So it makes sense for my usecase to use their role assignment.

The assigned groups are included in the access_token, so I figured it would be easiest for Keto to loop through it. 
Adding users and roles to Keto as an **additional** option would be perfectly acceptable to me.

**Comment by aeneasr** [comment]
> Which works, as long as you only have a single role. When you assign multiple roles it no longer works.

I think this needs clarification. Roles should definitely work in a m:n relationship with users. If that's not the case this is a bug (but I would be surprised if that's really the case).

Internally, we want to rework the Keto authorizer as a whole. We actually want to make it much more generic so that it works with OPA or any other type of policy engine as well. For that, we would use a similar Go Template pattern as we do for the header, cookie, and id_token mutator. This would also solve your use case as you could arbitrarily add or remove data from the payload. For example, you could write policies with conditions that a certain role must (or must'nt) be matched.

Do you think that would solve your issue in a more generic way? I would like to avoid having proprietary formats (defining roles in a JWT is nowhere standardized and thus proprietary) embedded in Oathkeeper, as it would diminish the applicability of the software as a whole and/or encourage/justify others to add more proprietary things to the stack.

**Comment by MichielVanderlee** [comment]
Keto can have multiple subjects defined per policy. But it'll only check a single subject against that list. 
So in Oathkeeper I can set the subject to check with via go templates. But I can only set a singular subject. 

Keto:
```
{
  "subjects": ["role1", "role2"],
  "resources": ["blog_posts:my-first-blog-post"],
  "actions": ["create"],
  "effect": "allow"
}
```
Oathkeeper
```
"authorizer": {
    "handler": "keto_engine_acp_ory",
    "config": {
      "required_action": "my:action:$1",
      "required_resource": "my:resource:$2:foo:$1"
      "subject": "{{ .Subject }}",
      "flavor": "exact"
    }
  }
```


I'd like to set multiple. 
If my token looks like this:
```
{ 
  "sub": "Bob",
  "roles": ["role2", "role4"]
}
```
Then I want it to match against any Keto policy that has either, "Bob", "role2", or "role4". 

So in order to be generic, I proposed that Oathkeeper's Keto config should allow for multiple values to be passed in to match against the policy.
i.e.: 
```
"subject": [
    "{{ .Subject }}",
    "{{ .Extra.roles }}"
]
```

**Comment by aeneasr** [comment]
> The assigned groups are included in the access_token, so I figured it would be easiest for Keto to loop through it.

Cycling back to that line for a second. While from a narrow point of view this absolutely is a acceptable use case, from a more elevated perspective this implies that roles aren't actually roles, they're another identity/principal embedded in the credential of the JSON Web Token.

In fact, this would open further questions such as: "What happens if role1 is allowed to do something, but role2 isn't?"

This is why we have avoided adding a multiple user check to Keto in the first place and instead opted to add roles which have a very clear and concise pattern of authorization (deny > allow > default (deny)).

I'm not sure if we would support multiple identities per token as a primary use case in Oathkeeper as the assumption of access control / authorization is always that a singular Identity is acting. 

While I do understand the wish to not duplicate things (e.g. AWS Cognito Groups with Roles from another system) you're actually already doing that by defining policies that apply to those very same roles. I agree that copying roles and users into another system is another level of complexity (keeping them in sync as they change over their lifetime) but to me (at least for now) it appears to be one of the few ways to always get consistent results over the span of a magnitude of use cases.

**However**, I think your specific use case is very interesting as it breaks out of the ORY-centric view and factors in other systems such as Cognito. We do however need to figure out a smart way to approach that where Oathkeeper would not be the decision point (your suggestions would imply that oathkeeper has to decide what to do if e.g. role 1 is allowed but role 2 isn't) but instead it would be the actual decision point (e.g. Keto).

I definitely have to do some thinking on this, it's a good use case and I would like to solve it, but I can't promise a swift resolution to this problem as it has a lot of implications and things to consider.

**Comment by MichielVanderlee** [comment]
> I definitely have to do some thinking on this, it's a good use case and I would like to solve it, but I can't promise a swift resolution to this problem as it has a lot of implications and things to consider.

That's all I ask :)

> In fact, this would open further questions such as: "What happens if role1 is allowed to do something, but role2 isn't?"

I'd expect an explicit 'deny' to take precedence. "(deny > allow > default (deny))" as you said.
Perhaps a good system to look at is AWS IAM. Where a user can be assigned multiple groups and each group multiple policies.

**Comment by aeneasr** [comment]
> I'd expect an explicit 'deny' to take precedence. "(deny > allow > default (deny))" as you said.
Perhaps a good system to look at is AWS IAM. Where a user can be assigned multiple groups and each group multiple policies.

The point is that this would make Oathkeeper essentially a decision point, which it is not (it's a enforcement point) as Oathkeeper would make the precedence decision over Keto's head. It would make the system more complicated to understand as implicit knowledge is required because both Keto (or any substitute like OPA) as well as Oathkeeper would have a rule that allows or denies something.

As Keto (or more general a decision point) is capable of and concerned with making this decision, it would be unwise to move that into another component as it would be a violation of the separation of concerns.

So basically, I think, we would need to make Keto (or any substitute) aware of other information pools (e.g. Cognito), instead of trying to fit this into Oathkeeper.

**Comment by aeneasr** [comment]
As Keto is now in a new version with new APIs this issue is now a wontfix, plus:

I am closing this issue as it has not received any engagement from the community or maintainers in a long time. That does not imply that the issue has no merit. If you feel strongly about this issue

- open a PR referencing and resolving the issue;
- leave a comment on it and discuss ideas how you could contribute towards resolving it;
- open a new issue with updated details and a plan on resolving the issue.

We are cleaning up issues every now and then, primarily to keep the 4000+ issues in our backlog in check and to [prevent](https://www.jeffgeerling.com/blog/2020/saying-no-burnout-open-source-maintainer) [maintainer](https://www.jeffgeerling.com/blog/2016/why-i-close-prs-oss-project-maintainer-notes) [burnout](https://docs.brew.sh/Maintainers-Avoiding-Burnout). Burnout in open source maintainership is a [widespread](https://opensource.guide/best-practices/#its-okay-to-hit-pause) and serious issue. It can lead to severe personal and health issues as well as [enabling catastrophic](https://haacked.com/archive/2019/05/28/maintainer-burnout/) [attack vectors](https://www.gradiant.org/en/blog/open-source-maintainer-burnout-as-an-attack-surface/).

Thank you to anyone who participated in the issue! 🙏✌️
